test/Interop/Cxx/value-witness-table/custom-destructors.swift

```diff
@@ -0,0 +1,140 @@
+// RUN: %target-run-simple-swift(-I %S/Inputs -Xfrontend -enable-cxx-interop)
+//
+// REQUIRES: executable_test
+
+import CustomDestructor
+import StdlibUnittest
+
+var CXXDestructorTestSuite = TestSuite("CXXDestructor")
+
+protocol InitWithPtr {
+  init(value: UnsafeMutablePointer<Int32>!)
+}
+
+extension HasUserProvidedDestructor : InitWithPtr { }
+
+protocol InitWithMember {
+  init(member: HasUserProvidedDestructor)
+}
+
+extension HasEmptyDestructorAndMemberWithUserDefinedConstructor
+  : InitWithMember { }
+
+// These are all maked as "@inline(never)" so that the destructor is invoked in
+// this function and not in the caller (after being inlined).
+@inline(never)
+func createTypeWithUserProvidedDestructor(_ ptr: UnsafeMutablePointer<Int32>) {
+  _ = HasUserProvidedDestructor(value: ptr)
+}
+
+@inline(never)
+func createTypeWithEmptyDestructorAndMemberWithUserDefinedConstructor(
+  _ ptr: UnsafeMutablePointer<Int32>
+) {
+  let obj = HasUserProvidedDestructor(value: ptr)
+  _ = HasEmptyDestructorAndMemberWithUserDefinedConstructor(
+    member: obj)
+}
+
+@inline(never)
+func createTypeWithNonTrivialImplicitDestructor(
+  _ ptr: UnsafeMutablePointer<Int32>
+) {
+  let obj = HasUserProvidedDestructor(value: ptr)
+  _ = HasNonTrivialImplicitDestructor(member: obj)
+}
+
+@inline(never)
+func createTypeWithNonTrivialDefaultDestructor(
+  _ ptr: UnsafeMutablePointer<Int32>
+) {
+  let obj = HasUserProvidedDestructor(value: ptr)
+  _ = HasNonTrivialDefaultedDestructor(member: obj)
+}
+
+@inline(never)
+func createTypeWithGeneric<T : InitWithPtr>(
+  _ ptr: UnsafeMutablePointer<Int32>,
+  type: T.Type
+) {
+  _ = T(value: ptr)
+}
+
+@inline(never)
+func createTypeWithProtocol(
+  _ ptr: UnsafeMutablePointer<Int32>,
+  type: InitWithPtr.Type
+) {
+  _ = type.self.init(value: ptr)
+}
+
+@inline(never)
+func createTypeWithProtocol(
+  _ ptr: UnsafeMutablePointer<Int32>,
+  type: InitWithPtr.Type,
+  holder: InitWithMember.Type
+) {
+  let obj = type.self.init(value: ptr)
+  _ = holder.self.init(member: obj as! HasUserProvidedDestructor)
+}
+
+CXXDestructorTestSuite.test("Basic object with destructor") {
+  let ptr = UnsafeMutablePointer<Int32>.allocate(capacity: 1)
+  ptr.pointee = 0
+  createTypeWithUserProvidedDestructor(ptr)
+  expectEqual(ptr.pointee, 42)
+  ptr.deallocate()
+}
```

Comment:
```suggestion
CXXDestructorTestSuite.test("Basic object with destructor") {
  var value: Int32 = 0
  createTypeWithUserProvidedDestructor(&valueptr)
  expectEqual(value, 42)
}
```
Please also apply in other test cases in this file.